src/theory/strings/rewrites

```diff
@@ -189,6 +215,11 @@
   (str.substr (str.++ s1 s2 s3) n m)
   (str.substr (str.++ s2 s3) (- n (str.len s1)) m))
 
+(define-cond-rule str-substr-replace ((s ?Seq) (t ?Seq) (r ?Seq) (n Int) (m Int))
```

Comment:
I don't quite get this one:


`(set-logic QF_S)

(declare-fun s () String)
(declare-fun t () String)
(declare-fun r () String)
(declare-fun LHS () String)
(declare-fun RHS () String)

(declare-fun n () Int)
(declare-fun m () Int)

(assert (= s "abcde"))
(assert (= t "de"))
(assert (= r "yy"))
(assert (= n 0))
(assert (= m 4))
(assert (= (str.len t) (str.len r)))

(assert (= LHS (str.substr (str.replace s t r) n m)))
(assert (= RHS (str.replace (str.substr s n m) t r)))

(check-sat)
(get-model)`

gives me the model:

`(define-fun LHS () String "abcy")
(define-fun RHS () String "abcd")`

which means that a match "abcy" could wrongly be rewritten as "abcd" or am I misinterpreting it?

Reply:
Great catch, indeed this rule is missing a condition that the length of t and r is one.

This is based off https://github.com/cvc5/cvc5/blob/main/src/theory/strings/sequences_rewriter.cpp#L2606, which has the additional restriction that n=0, which I believe is also necessary.